models/cm_compartmentcurrents.cpp

```diff
@@ -0,0 +1,510 @@
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with NEST.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+#include "cm_compartmentcurrents.h"
+
+
+// Na channel //////////////////////////////////////////////////////////////////
+nest::Na::Na()
+  : m_Na_( 0.0 )
+  , h_Na_( 0.0 )
+  , gbar_Na_( 0.0 )
+  , e_Na_( 50. )
+{
+}
+nest::Na::Na( const DictionaryDatum& channel_params )
+  : m_Na_( 0.0 )
+  , h_Na_( 0.0 )
+  , gbar_Na_( 0.0 )
+  , e_Na_( 50. )
+{
+  // update sodium channel parameters
+  if ( channel_params->known( "gbar_Na" ) )
+  {
+    gbar_Na_ = getValue< double >( channel_params, "gbar_Na" );
+  }
+  if ( channel_params->known( "e_Na" ) )
+  {
+    e_Na_ = getValue< double >( channel_params, "e_Na" );
+  }
+}
+
+void
+nest::Na::append_recordables( std::map< Name, double* >* recordables, const long compartment_idx )
+{
+  ( *recordables )[ Name( "m_Na_" + std::to_string( compartment_idx ) ) ] = &m_Na_;
+  ( *recordables )[ Name( "h_Na_" + std::to_string( compartment_idx ) ) ] = &h_Na_;
+}
+
+std::pair< double, double >
+nest::Na::f_numstep( const double v_comp, const double dt )
+{
+  double g_val = 0., i_val = 0.;
+
+  if ( gbar_Na_ > 1e-9 )
+  {
+    // auxiliary variables for 'm'
+    double a__times__v_comp__plus__b = 0.182 * v_comp + 6.3723659999999995;
+    double m_a__times__v_comp__minus_b = -0.124 * v_comp - 4.3416119999999996;
+    double one__minus__a__times__exp__mb__times__v_comp =
+      1.0 - 0.020438532058318047 * std::exp( -0.1111111111111111 * v_comp );
+    double one__minus__c__times__exp__d__times__v_comp =
+      1.0 - 48.927192870146527 * std::exp( 0.1111111111111111 * v_comp );
+
+    // activation and timescale of state variable 'm'
+    double m_inf_Na = a__times__v_comp__plus__b
+      / ( one__minus__a__times__exp__mb__times__v_comp
+        * ( m_a__times__v_comp__minus_b / one__minus__c__times__exp__d__times__v_comp
+          + a__times__v_comp__plus__b / one__minus__a__times__exp__mb__times__v_comp ) );
+    double tau_m_Na = 0.3115264797507788
+      / ( m_a__times__v_comp__minus_b / one__minus__c__times__exp__d__times__v_comp
+        + a__times__v_comp__plus__b / one__minus__a__times__exp__mb__times__v_comp );
+
+    // auxiliary variables for 'h'
+    double v_comp__div__5 = 0.20000000000000001 * v_comp;
+
+    // activation and timescale of state variable 'h'
+    double h_inf_Na = 1.0 / ( std::exp( 0.16129032258064516 * v_comp + 10.483870967741936 ) + 1.0 );
+    double tau_h_Na = 0.3115264797507788
+      / ( ( -0.0091000000000000004 * v_comp - 0.68261830000000012 )
+          / ( 1.0 - 3277527.8765015295 * std::exp( v_comp__div__5 ) )
+        + ( 0.024 * v_comp + 1.200312 ) / ( 1.0 - 4.5282043263959816e-5 * exp( -v_comp__div__5 ) ) );
```

Comment:
Could the origin of these numbers be explained somehow (e.g., by a reference) or be calculated in the code here from more obvious quantities? Many of the values seem to be the result of some external calculation where they were subject to some rounding (e.g. -0.0091000000000000004).

Reply:
I added refs to the original files from where these number come. Also made the implementation clearer, now the computation starts (mostly) from the base parameters as they occur in the original models.

Reply:
Much nicer indeed. Thanks!